code/datums/extensions/exonet.dm

```diff
@@ -0,0 +1,161 @@
+	if(!new_ennid)
+		return
+	var/new_key = sanitize(input(user, "Enter exonet keypass or leave blank if none:", "Change Key"))
+
+	var/found = FALSE
+	for(var/datum/exonet/network in get_nearby_networks(netspeed))
+		if(network.ennid == new_ennid)
+			// We found our network.
+			result["ennid"] = new_ennid
+			result["key"] = new_key
+			found = TRUE
+	if(!found)
+		var/network_list = list()
+		for(var/datum/exonet/network in get_nearby_networks(netspeed))
+			network_list |= network.ennid
+		if(!length(network_list))
+			network_list |= "None"
+		result["error"] = "Unable to find network with ennid '[new_ennid]'. Available networks: [jointext(network_list, ", ")]."
+	return result
+
+/datum/extension/exonet_device/proc/do_change_net_tag(var/mob/user)
+	var/list/result = list()
+
+	var/new_tag = sanitize(input(user, "Enter exonet network tag or leave blank to cancel:", "Change Network Tag"))
+	if(!new_tag)
+		return
+
+	var/datum/exonet/network = get_local_network()
+	if(!network)
+		result["error"] = "Cannot change network tag while disconnected from network."
+		return result
+	for(var/network_device in network.network_devices)
+		if(lowertext(get_network_tag(network_device)) == lowertext(new_tag))
+			result["error"] = "Network tags must be unique."
+			return result
+	result["net_tag"] = new_tag
+	return result
+
+/datum/extension/exonet_device/proc/do_change_key(var/mob/user)
+	var/list/result = list()
+	var/new_key = sanitize(input(user, "Enter exonet keypass or leave blank if none:", "Change Key"))
+	
+	if(!new_key)
+		return
+	result["key"] = new_key
+	return result		
+
+/datum/extension/exonet_device/proc/connect_network(var/mob/user, var/new_ennid, var/nic_netspeed, var/keydata)
+	if(ennid == new_ennid)
+		return "\The [holder] is already part of the '[ennid]' network."
+
+	if(ennid)
+		var/disconnect_result = disconnect_network(user)
+		if(disconnect_result)
+			return disconnect_result // There was a problem.
+
+	var/datum/exonet/exonet = GLOB.exonets[new_ennid]
+	if(!exonet)
+		return "Error encountered when trying to register \the [holder] to the '[new_ennid]' network."
+	else
+		if(!exonet.router)
+			return "Error encountered when trying to register \the [holder] to the '[new_ennid]' network."
+		if(exonet.router.keydata != keydata)
+			return "Invalid key. Unable to connect to network."
+		exonet.add_device(holder, keydata)
+		ennid = new_ennid
+		netspeed = nic_netspeed
+	return FALSE // This is a success.
+
+/datum/extension/exonet_device/proc/disconnect_network(var/mob/user)
+	if(!ennid)
+		return
+
+	var/datum/exonet/old_exonet = GLOB.exonets[ennid]
+	if(old_exonet)
+		old_exonet.remove_device(holder)
+		ennid = null
+	return FALSE // This is a success.
+
+/datum/extension/exonet_device/proc/broadcast_network(var/b_ennid, var/key)
+	// Broadcasts an ENNID. If the network doesn't exist, it will create it.
+	// If the network does exist, this will attempt to be added as a relay.
```

Comment:
`will attempt to be added as a relay` seems to be just added like a normal device tho?

Reply:
add_device in exonet datum does the check to add it as a broadcaster.